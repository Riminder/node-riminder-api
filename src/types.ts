export interface ProfilesOptions {
  source_ids: Array<string>;
  date_start: Date;
  date_end: Date;
  page: number;
  seniority?: string;
  filter_id?: string;
  filter_reference?: string;
  stage?: string;
  rating?: string;
  limit?: number;
  sort_by?: string;
  order_by?: string;
}

interface ProfileOptionId {
  source_id: string;
  profile_id: string;
}

interface ProfileOptionReference {
  source_id: string;
  profile_reference: string;
}

export type ProfileOptionIdOrReference = ProfileOptionId | ProfileOptionReference;

interface TrainingMetadataId {
  filter_id: string;
  stage: string;
  stage_timestamp: Date;
  rating: number;
  rating_timestamp: Date;
}

interface TrainingMetadataReference {
  filter_reference: string;
  stage: string;
  stage_timestamp: Date;
  rating: number;
  rating_timestamp: Date;
}

export interface ProfileUpload {
  source_id: string;
  profile_reference: string;
  timestamp_reception: Date;
  training_metadata: Array<TrainingMetadataId | TrainingMetadataReference>;
}

interface StagePatchProfileIdFilterId {
  source_id: string;
  profile_id: string;
  filter_id: string;
  stage: string | null;
}

interface StagePatchProfileIdFilterReference {
  source_id: string;
  profile_id: string;
  filter_reference: string;
  stage: string | null;
}

interface StagePatchProfileReferenceFilterId {
  source_id: string;
  profile_reference: string;
  filter_id: string;
  stage: string | null;
}

interface StagePatchProfileReferenceFilterReference {
  source_id: string;
  profile_reference: string;
  filter_reference: string;
  stage: string | null;
}

export type StagePatch =
  StagePatchProfileIdFilterId |
  StagePatchProfileIdFilterReference |
  StagePatchProfileReferenceFilterId |
  StagePatchProfileReferenceFilterReference;