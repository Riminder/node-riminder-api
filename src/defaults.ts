let defaults: any = {
  API_URL: "https://www.riminder.net/sf/public/api/v1.0",
  API_SECRET: null,
  API_Key: null
};

export default defaults;