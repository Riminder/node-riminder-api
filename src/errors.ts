import { RiminderAPIResponse } from "./types";

export class APIError extends Error {
    response: RiminderAPIResponse;
    constructor(message: string, response: RiminderAPIResponse)Â {
        super(message);
        this.response = response;
    }
}