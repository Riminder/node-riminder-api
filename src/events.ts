export default [
  "profile.parse.success",
  "profile.parse.error",
  "profile.score.success",
  "profile.score.error",
  "filter.train.start",
  "filter.train.success",
  "filter.train.error",
  "filter.score.start",
  "filter.score.success",
  "filter.score.error"
];