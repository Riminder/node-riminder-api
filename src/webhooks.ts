import Events from "./events";
import * as util from "tweetnacl-util";
import * as sha256 from "fast-sha256";

export namespace Webhooks {
  export interface ResponseBase {
    type: string;
    message: string;
  }
  export interface ProfileParseResponse extends ResponseBase {
    profile: Profile;
  }

  export interface ProfileScoreResponse extends ResponseBase {
    profile: Profile;
    filter: Filter;
    score: number;
  }

  export interface FilterTrainResponse extends ResponseBase {
    filter: Filter;
  }

  export interface FilterScoreResponse extends ResponseBase {
    filter: Filter;
  }

  export type Response = ProfileParseResponse | ProfileScoreResponse | FilterTrainResponse |Â FilterScoreResponse;

  export interface Profile {
    profile_id: string;
    profile_reference: string;
  }

  export interface Filter {
    filter_id: string;
    filter_reference: string;
  }

  export type EventCallbackMap = Map<string, (data: Webhooks.Response) => any>;
}

export class Webhooks {
  private webhookSecretKey: string;
  binding: Webhooks.EventCallbackMap;

  constructor(secretKey: string) {
    if (!secretKey) {
      throw new Error("The webhook secret key must be specified");
    }

    this.webhookSecretKey = secretKey;
    this.binding = new Map<string, (data: Webhooks.Response) => any>();
  }

  handleWebhook(headers: any): () => void {
    return () => {
      if (!headers["HTTP-RIMINDER-SIGNATURE"]) {
        throw new Error("The signature is missing from the headers");
      }

      const [encodedSignature, encodedPayload] = headers["HTTP-RIMINDER-SIGNATURE"].split(".");
      const expectedSignature = util.encodeBase64(sha256.hmac(util.decodeUTF8(this.webhookSecretKey), util.decodeUTF8(encodedPayload)));

      if (encodedSignature !== expectedSignature) {
        throw new Error("The signature is invalid");
      }

      const payload: Webhooks.Response = JSON.parse(util.encodeUTF8(util.decodeBase64(encodedPayload)));

      if (Events.indexOf(payload.type) < 0) {
        throw new Error(`Unknown event: ${payload.type}`);
      }
      this._callBinding(payload);
    };
  }

  on(event: string, callback: (data: Webhooks.Response) => any) {
    if (Events.indexOf(event) < 0) {
      throw new Error("This event doesn't exist");
    }

    if (this.binding.has(event)) {
      throw new Error("This callback already has been declared");
    }

    this.binding.set(event, callback);

    return this;
  }

  private _callBinding(payload: Webhooks.Response): void {
    if (this.binding.has(payload.type)) {
      this.binding.get(payload.type)(payload);
    }
  }
}