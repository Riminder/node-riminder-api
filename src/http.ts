import "whatwg-fetch";
import "./defaults";
import defaults from "./defaults";
import { Riminder } from "./index";

declare interface RiminderAPIResponse {
  code: number;
  message: string;
  data?: any;
}

export const httpRequest = (url: string, options?: any) => {
  let headers = {
    "X-API-Key": Riminder._instance.API_Key || defaults.API_Key,
  };

  let opts = {
    headers,
    credentials: "include",
    ...options
  };

  return fetch(url, opts)
    .then(successHandler, errorHandler)
    .then((json: RiminderAPIResponse) => json.data);
};

const successHandler = (response: Response) => {
  if (response.status === 200 || response.status === 201) {
    return response.json();
  }
  console.error(response);
};

const errorHandler = (err: any) => {
  let error = new Error(err.message);
  (<any>error).response = err;
  throw error;
};