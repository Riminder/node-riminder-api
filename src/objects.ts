import { create, getOne, getList, generateURLParams } from "./utils";
import defaults from "./defaults";
import { httpRequest } from "./http";

export interface RiminderObjectsOptions {
  where?: any;
  page?: number;
  itemsPerPage?: number;
}

export default {
  getSources: (options?: RiminderObjectsOptions) => {
    return getList("sources", options);
  },
  getSource: (id: string) => {
    return getOne("source", id);
  },
  getProfiles: (options?: RiminderObjectsOptions) => {
    return getList("profiles", options);
  },
  getProfile: (id: string, sourceID: string) => {
    let url = `${defaults.API_URL}/profile/${id}?source_id=${sourceID}`;
    return httpRequest(url);
  },
  createResumeForProfile: (profileID: string, sourceID: string, file: File) => {

  },
  getProfileDocuments: (id: string, sourceID: string) => {
    let url = `${defaults.API_URL}/profile/${id}/documents?source_id=${sourceID}`;
    return httpRequest(url);
  },
  getProfileExtractions: (id: string, sourceID: string) => {
    let url = `${defaults.API_URL}/profile/${id}/extractions?source_id=${sourceID}`;
    return httpRequest(url);
  },
  getProfileJobs: (id: string, sourceID: string) => {
    let url = `${defaults.API_URL}/profile/${id}/jobs?source_id=${sourceID}`;
    return httpRequest(url);
  },
  updateProfileStage: (id: string, sourceID: string, jobID: string, stage: string) => {
    let url = `${defaults.API_URL}/profile/${id}/stage`;
    let body = {
      source_id: sourceID,
      job_id: jobID,
      stage: stage,
    };

    return httpRequest(url, { body });
  },
  updateProfileRating: (id: string, sourceID: string, jobID: string, rating: number) => {
    let url = `${defaults.API_URL}/profile/${id}/rating`;
    let body = {
      source_id: sourceID,
      job_id: jobID,
      rating: rating,
    };

    return httpRequest(url, { body });
  }
};
