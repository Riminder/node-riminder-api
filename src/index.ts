import { httpRequest } from "./http";
import { create, getOne, getList } from "./utils";
import objects from "./objects";

export interface RiminderOptions {
  API_Key: string;
  API_Secret?: string;
}

export class Riminder {
  static _instance: Riminder;
  public API_Key: string;
  public objects: any;
  public webhooks: any;
  constructor(options: RiminderOptions) {
    if (Riminder._instance) {
      let error = new Error("You can not instanciate more than one instance of Riminder SDK");
      throw error;
    }

    if (!options.API_Key) {
      let error = new Error("No API Key was supplied for Riminder SDK");
      throw error;
    }

    this.API_Key = options.API_Key;
    this._init();
    Riminder._instance = this;
  }

  private _init() {
    this.objects = objects;
    this.webhooks = {};
  }

}