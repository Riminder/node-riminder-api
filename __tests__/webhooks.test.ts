import Riminder = require("../src/index");
import { Webhooks } from "../src/webhooks";
import * as util from "tweetnacl-util";
import * as sha256 from "fast-sha256";

let app: Riminder;
const API_Key = "api_key";
const Webhooks_Key = "webhooks_key";

function generateSignature(key: string, event: string) {
    const data: Webhooks.ProfileParseResponse = {
        type: event,
        message: "The parsing succeeded",
        profile: {
            profile_id: "id",
            profile_reference: "reference"
        }
    };

    const JSONPayload = JSON.stringify(data);
    const base64Payload = util.encodeBase64(util.decodeUTF8(JSONPayload));

    const signature = sha256.hmac(util.decodeUTF8(key), util.decodeUTF8(base64Payload));
    const base64Signature = util.encodeBase64(signature);

    return `${base64Signature}.${base64Payload}`;
}

const callbackMock = jest.fn();

describe("Webhooks tests",  () => {
    describe("Webhook check", () => {
        test("It should check if the webhook is correctly set up", () => {
            app = new Riminder({ API_Key, Webhooks_Key });
            app.webhooks.check()
                .then((response) => {
                    expect(response).toMatchSnapshot();
                });
        });
    });

    describe("Webhook creation", () => {
        test("It should not create webhooks if no key is given", () => {
            app = new Riminder({ API_Key });
            expect(app.webhooks).toBeUndefined();
        });

        test("It should create webhooks if the key is given", () => {
            app = new Riminder({ API_Key, Webhooks_Key });
            expect(app.webhooks).toBeDefined();
        });

        test("It should throw an error if we try to create webhooks without a key", () => {
            expect(() => {
                const webhooks = new Webhooks(null);
            }).toThrowError("The webhook secret key must be specified");
        });
    });

    describe("Webhooks usage", () => {
        describe("Webhooks binding", () => {
            beforeEach(() => {
                app = new Riminder({ API_Key, Webhooks_Key });
            });

            test("It should throw an error if we try to bind an inexistant event", (done) => {
                expect(() => {
                    app.webhooks.on("not.an.event", (data: Webhooks.Response, type: string) => {});
                }).toThrowError("This event doesn't exist");
                done();
            });

            test("It should throw an error if we try to bind two functions to the same event", (done) => {
                expect(() => {
                    app.webhooks
                        .on("profile.parse.success", (data: Webhooks.Response, type: string) => {})
                        .on("profile.parse.success", (data: Webhooks.Response, type: string) => {});
                }).toThrowError("This callback already has been declared");
                done();
            });

            test("It should bind the event correctly", (done) => {
                expect(app.webhooks.on("profile.parse.success", (data: Webhooks.Response, type: string) => {
                    console.log(type);
                    return 42;
                })).toBeInstanceOf(Webhooks);
                expect(app.webhooks.binding.has("profile.parse.success")).toBeTruthy();
                expect(app.webhooks.binding.get("profile.parse.success")({} as Webhooks.Response, "test")).toBe(42);
                done();
            });
        });

        describe("Webhooks call", () => {
            beforeAll(() => {
                app = new Riminder({ API_Key, Webhooks_Key });
                app.webhooks.on("profile.parse.success", callbackMock);
            });

            test("It should throw an error if the header is not given", () => {
                expect(app.webhooks.handle({})).toThrowError("The signature is missing from the headers");
            });

            test("It should throw an error if the signature is invalid", () => {
                const signature = generateSignature("wrong_key", "profile.parse.success");
                const headers = {
                    "HTTP-RIMINDER-SIGNATURE": signature
                };
                expect(app.webhooks.handle(headers)).toThrowError("The signature is invalid");
            });

            test("It should throw an error if the event is unknown", () => {
                const signature = generateSignature(Webhooks_Key, "unknwown.event");
                const headers = {
                    "HTTP-RIMINDER-SIGNATURE": signature
                };
                expect(app.webhooks.handle(headers)).toThrowError("Unknown event: unknwown.event");
            });

            test("It should call the callbacj function", () => {
                const signature = generateSignature(Webhooks_Key, "profile.parse.error");
                const headers = {
                    "HTTP-RIMINDER-SIGNATURE": signature
                };
                app.webhooks.on("profile.parse.error", callbackMock);
                app.webhooks.handle(headers)();
                expect(callbackMock).toHaveBeenCalledTimes(1);
            });
        });
    });
});