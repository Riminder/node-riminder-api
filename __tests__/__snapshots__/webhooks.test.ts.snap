// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webhooks tests Webhook check It should check if the webhook is correctly set up 1`] = `
Object {
  "options": Object {
    "body": null,
    "headers": Object {
      "X-API-Key": "api_key",
    },
    "method": "POST",
  },
  "url": Url {
    "auth": null,
    "hash": null,
    "host": "www.riminder.net",
    "hostname": "www.riminder.net",
    "href": "https://www.riminder.net/sf/public/api/v1.0/webhook/check",
    "path": "/sf/public/api/v1.0/webhook/check",
    "pathname": "/sf/public/api/v1.0/webhook/check",
    "port": null,
    "protocol": "https:",
    "query": null,
    "search": null,
    "slashes": true,
  },
}
`;
